audiostream/src/ni/media/audio/source/gstreamer_file_source.cpp

```diff
@@ -0,0 +1,336 @@
+
+pcm::runtime_format gstreamer_file_source::create_runtime_format( GstStructure* caps_struct )
+{
+    const gchar*     format      = gst_structure_get_string( caps_struct, "format" );
+    pcm::number_type number_type = gst_format_char_to_number_type( format[0] );
+    auto             srcDepth    = std::atol( format + 1 );
+    auto             endian      = ( strcmp( format, "BE" ) == 0 ) ? pcm::big_endian : pcm::little_endian;
+    return pcm::runtime_format( number_type, srcDepth, endian );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+pcm::number_type gstreamer_file_source::gst_format_char_to_number_type( const gchar format )
+{
+    if ( format == 'U' )
+        return pcm::unsigned_integer;
+    else if ( format == 'F' )
+        return pcm::floating_point;
+
+    return pcm::signed_integer;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::onPadAdded( GstElement* element, GstPad* pad, GstElement* sink )
+{
+    tGstPtr<GstPad> sinkpad( gst_element_get_static_pad( sink, "sink" ), gst_object_unref );
+    gst_pad_link( pad, sinkpad.get() );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+std::streampos gstreamer_file_source::seek( offset_type off, BOOST_IOS::seekdir way )
+{
+    assert( 0 == off % m_info.bytes_per_frame() );
+
+    int64_t newPosition = 0;
+
+    if ( way == BOOST_IOS::seekdir::_S_beg )
+    {
+        newPosition = off;
+    }
+    else if ( way == BOOST_IOS::seekdir::_S_cur )
+    {
+        newPosition = m_position + off;
+    }
+    else if ( way == BOOST_IOS::seekdir::_S_end )
+    {
+        int64_t end = 0;
+        if ( !gst_element_query_duration( m_pipeline.get(), GST_FORMAT_BYTES, &end ) )
+        {
+            throw std::runtime_error(
+                "gstreamer_file_source: seeking from end of file impossible - could not query duration" );
+        }
+        newPosition = end + off;
+    }
+
+    if ( m_position != newPosition )
+    {
+        m_position = newPosition;
+        m_ring_buffer->flush();
+
+        if ( !gst_element_seek_simple( m_pipeline.get(),
+                                       GST_FORMAT_BYTES,
+                                       ( GstSeekFlags )( GST_SEEK_FLAG_FLUSH | GST_SEEK_FLAG_ACCURATE ),
+                                       m_position ) )
+        {
+            throw std::runtime_error( "gstreamer_file_source: seeking failed" );
+        }
+    }
+
+    return m_position;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+std::streamsize gstreamer_file_source::read( char* dst, std::streamsize numBytesRequested )
+{
+    return recursive_read( dst, numBytesRequested );
```

Comment:
Is there a reason why you chose to loop recursively instead of iteratively in the read function?

Reply:
No, there is no special reason, except: I always do it like this. Sometimes, as in RingBuffer::push/pull, it improves debug-ability as you can inspect all the 'const' variables on the according stack frames. 
However, I do not insist having recursive algorithms if you guys don't like it.

Reply:
For sure, was just curious!  We have been mainly using iterative read/write loops in our sources and sinks except for MediaFoundation.  I don't feel strongly towards one or the other, and @marcrambo doesn't as well.